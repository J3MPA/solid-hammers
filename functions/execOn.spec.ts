import execOn from './execOn'
describe('execOn', () => {
  describe('Happy path', () => {
    it('should execute fn on nth time of execution', () => {
      const mockFn = jest.fn((s: string, n: number) => ({ s, n }))
      const fn = execOn(3, mockFn)

      fn('first', 1)
      expect(mockFn).toHaveBeenCalledTimes(0)

      fn('second', 2)
      expect(mockFn).toHaveBeenCalledTimes(0)

      fn('third', 3)
      expect(mockFn).toHaveBeenCalledTimes(1)
      expect(mockFn).toHaveBeenCalledWith('third', 3)
      expect(mockFn).toHaveReturnedTimes(1)
      expect(mockFn).toHaveReturnedWith({ s: 'third', n: 3 })
    })
  })

  describe('Sad path', () => {
    it('should throw TypeError if nth time is not a real number', () => {
      const mockFn = jest.fn()
      expect(() => execOn(NaN, mockFn)).toThrowError(TypeError)
      expect(() => execOn(NaN, mockFn)).toThrowErrorMatchingSnapshot()
      expect(() => execOn(Infinity, mockFn)).toThrowError(TypeError)
      expect(() => execOn(Infinity, mockFn)).toThrowErrorMatchingSnapshot()
      expect(() => execOn(null as any, mockFn)).toThrowError(TypeError)
      expect(() => execOn(null as any, mockFn)).toThrowErrorMatchingSnapshot()
      expect(() => execOn(undefined as any, mockFn)).toThrowError(TypeError)
      expect(() =>
        execOn(undefined as any, mockFn)
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
