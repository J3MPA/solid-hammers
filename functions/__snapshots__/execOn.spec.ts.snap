// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execOn Sad path should throw TypeError if nth time is not a real number 1`] = `"nthTime must be a real number"`;

exports[`execOn Sad path should throw TypeError if nth time is not a real number 2`] = `"nthTime must be a real number"`;

exports[`execOn Sad path should throw TypeError if nth time is not a real number 3`] = `"nthTime must be a real number"`;

exports[`execOn Sad path should throw TypeError if nth time is not a real number 4`] = `"nthTime must be a real number"`;
